### DISCUSSION TITLE
Multilingual-vits /// Blog post

### DISCUSSION THREAD

**Discussion by vince62s** [coqui-ai/TTS#1058] [General Q&A] [answered]
Hello,
Happy new year and congrats for the nice work.

I am curious to know you came up to train incrementally as said in the blog post.

Does this mean that we need to use the recipe several times with a train from a checkpoint at each increment ?

Does the model train properly if we train everything at the same time ?

Thanks.

**Comment by vince62s** [comment]
btw, the blog post says
Dataset paragraph: it uses VCTK + LibriTTS
Training paragraph: it starts with a pretrained no LJSpeech (1M steps) then VCTK 200K steps
Does it use LibriTTS at all ?

**Reply by erogol** [reply]
Sorry for the confusion. I think we forgot to mention the LibriTTS in the blog post.

**Comment by vince62s** [comment]
well reading the paper https://arxiv.org/abs/2112.02418 gives some answers, but replicability does not seem so straight forward to accomplish.
I'll give a try.

**Reply by erogol** [reply]
Let us know how it went.

**Reply by WeberJulian** [reply]
We'll publish a recipe soon for YourTTS, so if you struggle it might be worth waiting for that :)
In the meantime, if you're stuck somewhere you can ask for help here!

**Reply by vince62s** [reply]
At this point, I am trying to make sure I fully understand the current recipe under multilingual/vits
Actually I got two questions:
1) I did not find any setting to enable weighting a dataset versus others
2) For finetuning same question but also for speaker finetuning, I saw some kind of "def get_speaker_weighted_sampler(items: list):"
function but I'm not sure if I need to dig in this direction to fine tune an existing model on a specific speaker.

**Reply by WeberJulian** [reply]
1. The recipe under multilingual/vits isn't the one used for YourTTS. It doesn't use d_vectors but internal embeddings.
2. In the paper, we create a new language en-sa for example, and set the sa dataset to this language -> 1/4 of the batch

**Reply by hemath1001** [reply]
> We'll publish a recipe soon for YourTTS, so if you struggle it might be worth waiting for that :) In the meantime, if you're stuck somewhere you can ask for help here!

Hi WeberJulian! Will training and fine-tuning code be published soon? I would be very grateful if you could tell me an approximate time~ Have a nice day  :-D

**Reply by Ca-ressemble-a-du-fake** [reply]
This will be really helpful to have this recipe! I have been struggling for weeks to fine-tune the model with very bad results in the end. So looking forward to using this recipe!

**Reply by OmiWakode** [reply]
A similar request from my end! @WeberJulian. Can you tell us when are you planning to release the recipe for YourTTS model

**Comment by vince62s** [comment]
Hi,
I was able to start the mailabs training quite easily, still running.
I am starting to synthesize.
using tts (through python3 TTS/bin/synthesize.py) command line I can generate some wav files with existing speaker ids, in multiple languages. Works fine.

using the args --speaker_wav dummy.wav --encoder_path SE_checkpoint.pth.tar --encoder_config_path config_se.json

EDIT: damned .... the Speaker encoder checkpoint uses 512 dims vs the vits default config which is 256
Anyway to make a dim reduction of the SE ?

**Reply by erogol** [reply]
Our SE used in the paper uses 512 too. It should work just fine.

**Reply by OmiWakode** [reply]
@vince62s @erogol @WeberJulian 
I don't know if this is supposed to be asked here but I am kinda stuck and need some help.
1) Can we achieve Zero-Shot cloning with vits along with a separate speaker-encoder? I have my custom data set and want to replace the YourTTS model with my own vits model in the zero-shot cloning process (from the demo colab ). ?

2) Is it necessary to give a d_vector_dim in vits config in order to train the model on external speaker_embeddings. I used the same model for speaker embeddings from the YourTTS paper, I can't understand what's wrong. Also if I remove d_vector_dim, the model does work and starts training.
When I try to pass a d_vector_dim in vits args as 512, it throws an error :
(TypeError: conv1d() received an invalid combination of arguments - got (NoneType, Parameter, Parameter, tuple, tuple, tuple, int), but expected one of:
 * (Tensor input, Tensor weight, Tensor bias, tuple of ints stride, tuple of ints padding, tuple of ints dilation, int groups)
      didn't match because some of the arguments have invalid types.) 

I would like to get some clarity.

**Reply by WeberJulian** [reply]
Hey
1) Yes, that's what we did with YourTTS, not sure what you mean
2) Create an issue, it would be easier to help, with context and the whole trace

**Reply by OmiWakode** [reply]
I have created a new issue here [Error initializing d_vector_dim in VitsArgs](https://github.com/coqui-ai/TTS/issues/1454)
Thanks for replying

**Comment by vince62s** [comment]
I know it does, but I trained the mailabs vits with default config which is 256 ....
so I'll retrain.

btw, even without using a speaker encoder, the output is "ok" like clear but pace is very slow.

**Reply by WeberJulian** [reply]
If you are trying to reproduce results of the paper, I'll advise starting from the config that is downloaded when you do `tts --text "hello world!" --model_name "tts_models/multilingual/multi-dataset/your_tts" --speaker_wav "577_394_000000.wav" --language_idx "en"`

**Reply by Ca-ressemble-a-du-fake** [reply]
Where should the 577_394_000000.wav" be, in which dataset ? When running this I get ` No such file or directory: '577_394_000000.wav'`. Moreover are there news about the recipe soon to be published you mentioned earlier ?

**Reply by vince62s** [reply]
It's just an example. You need to read the code and the comments in the code to really understand how things work.

**Reply by Ca-ressemble-a-du-fake** [reply]
Oh ok! This is actually what the demo page does, where you input text, and a sample of your voice (--speaker_wav argument) and then it outputs the text said by your voice, right ? Do you think I would be able to fine tune this model with my own French dataset of 100 samples ? Because so far I fine-tuned the model from Exp3 and it does not sound like the target voice (from my dataset). But I guess my samples are overweight by the samples from MAILABS. Does this make sense,or am I completely off the track ?

**Reply by vince62s** [reply]
First if you want this model to be the starting point you need to understand how it has been trained (d_vectors, SCL, ...). Second there are several method to finetune. Some people will just use the new dataset (new = new voice only) but you take the risk to break things. Otherwise you need to find the proper weight vs the original data. Now what you seem to say is not finetuning their model (which in this case would mean to recompute all the d_vectors for their datasets). Anyway, only way is to try try try, but also to understand exactly what  you are doing. Enjoy.

**Reply by Ca-ressemble-a-du-fake** [reply]
Thank you @vince62s . There are so many parameters ! But I will look at the paper again and follow your advice to know how Exp 3 was originaly trained.

**Comment by vince62s** [comment]
@WeberJulian I am struggling a bit.
First trying to make sure I undertsand the code.
Internal embeddings (use_speaker_embedding=true) and d_vector (use_d_vector_file=true) are mutually exclusive, right ?
would it be possible to compute embeddings in the same way it is used with d_vectors so that we could mix both world ?
anyway beyond the scope of this thread.

I am now trying to train with d_vectors and using the d_vector_file speakers.json from the YourTTS config
I am hitting a wall in the dataset.py here:
https://github.com/coqui-ai/TTS/blob/main/TTS/tts/datasets/dataset.py#L418

it seems it's getting d_vectors from the dataset wavfiles instead of the speakers.json

So I changed it to:

            # get pre-computed d-vectors
            if self.d_vector_mapping is not None:
            #    wav_files_names = list(batch["wav_file_name"])
            #    d_vectors = [self.d_vector_mapping[w]["embedding"] for w in wav_files_names]
                d_vectors = [self.d_vector_mapping[random.choices(list(self.d_vector_mapping.keys()))[0]]["embedding"]]
            else:
                d_vectors = None

to randomly pick an embedding from the speakers.json file

Am I heading to the right direction ?


Side question: when looking inside the pretrained YourTTS model, I am seeing a "speaker_encoder"  layers config that I don't know where it's coming from used for.

**Reply by WeberJulian** [reply]
> would it be possible to compute embeddings in the same way it is used with d_vectors so that we could mix both world ?

I don't understand what you are trying to do, you mean concat the d_vector and internal embedding before feeding it to the model?

> to randomly pick an embedding from the speakers.json file

I don't get it either, you pick randomly d_vectors for training ?? Each d_vector is computed from an audio file, during training, you need to feed the model with the d_vector associated with GT.

> when looking inside the pretrained YourTTS model, I am seeing a "speaker_encoder" layers config that I don't know where it's coming from used for.

That's for SCL as explained in the paper, it's not necessary to get decent zs-TTS

**Comment by vince62s** [comment]
Forget about the first point for now.
For the second, I may be confused by the training procedure.
my understandging was
use_d_vector_file=true
d_vector_file=speakers.json
as in the config.json file from the YourTTS downloaded files.
In this scenario, I thought at training the speaker_embeddings were picked from that speakers.json file (randomly ?)
But your above answer suggests it's not the case.
If the speaker_embedding is the one from the GT wav file, is it computed on the fly ? if so where ? if not do we need to pre-compute ALL embeddings ?
I hope my question is clearer.

**Reply by WeberJulian** [reply]
You need to precompute the d_vectors with this script: https://github.com/coqui-ai/TTS/blob/main/TTS/bin/compute_embeddings.py and set d_vector_file with the resulting file.

**Comment by vince62s** [comment]
Got it. Another question, somehow linked to my first point.
When you train using d_vectors + pre-computed d_vect_file = speakers.json
it recreates in the output_path a speakers.json only with IDs (same format as when using internal embeddings) see here: https://github.com/coqui-ai/TTS/blob/main/TTS/tts/models/base_tts.py#L444
But unless I'm wrong you can infer with just an ID, you need the full embedding if you don't pass a wav file as input. Am I correct ?
if so, unsure why the speakers.json file is recreated in this context.
Hope this is clear.

**Reply by WeberJulian** [reply]
Yeah, that's a path of improvement for us, the generated file should be speaker_ids.json and not speakers.json that can be confused with the d_vectors file. This behavior is gonna change soon anyway since everything is gonna be contained in checkpoint keys.

**Comment by vince62s** [comment]
@WeberJulian more questions for you.
Beside the fact that you guys started from an LJspeech model trained during 1M steps (huge for 23 hours, no ?), paper refers to additional steps from 140k to 200k depending each added dataset. Did it really improve after 50-60K steps in each experiment ?
Also, how important is the number of speakers to achieve ZS-tts (eg is VCTK really more suitable than Mailabs - in terms of number of speakers, not languages of course).

**Comment by vince62s** [comment]
More experiments.
- added LibriTTS (1151 speakers)
- added more text encoder layers (10 same as in the paper).
- tried to add SCL.
No luck. The model is very good and also good at similarity with seen speakers but ZS-TTS is pretty bad in terms of similarity.
I don't think I missed a key point of the paper. Of course dataset is not the same since I am taking mailabs + VCTK + LibriTTS totalling 7 languages.

**Comment by vince62s** [comment]
What is the most relevant loss graph to look at ?
Surprinsgly the best model at 200K steps remained very long the best until 480K steps. So not easy to decide when to stop training.

**Comment by vince62s** [comment, marked as answer]
Bottom line there is no magic. ZS-TTS works for some voices not for others, even after 1M steps. When you finetune, it gets really good.
best is to check SECS with this https://github.com/resemble-ai/Resemblyzer as in the paper.

**Reply by OmiWakode** [reply]
Hi, you seem to have a greater understanding for training YourTTS. Can I know which model should I use for computing external speaker embeddings? I want to train a custom ZS-TTS on my own data set.

**Reply by WeberJulian** [reply]
Hey, you can find the speaker encoder model here https://github.com/edresson/yourtts#checkpoints

**Reply by e0xextazy** [reply]
@WeberJulian Which of the checkpoints would you advise using for a separate male voice and for a separate female voice?

**Reply by WeberJulian** [reply]
For fine-tuning we only experimented with the checkpoint from Exp. 4

**Reply by e0xextazy** [reply]
@WeberJulian Could you look at my charts and tell me if they are behaving normally or not? I'm confused by the fact that loss1 is going up and loss0 is going down
<img width="2054" alt="Снимок экрана 2022-03-15 в 18 51 11" src="https://user-images.githubusercontent.com/49094658/158418139-1bcf7b32-3f58-497b-bb43-792a0b808184.png">
I took my own vote of 100 examples and added more LibriTTS 100 and 360. I continued with experiment number 4, lowering the learning rate a bit and leaving the config as standard

**Comment by e0xextazy** [comment]
@vince62s Hi, having read the guides, discussions, articles and issues still left little understanding of how to fine tune the model to the target voice. I see that you have a very good understanding, so could you please give what steps you think are necessary to achieve a good result. The main problem is with understanding the config, as there are so many parameters. Thank you very much!

**Comment by vince62s** [comment]
I used the vits multilingual recipe with M Ailabs + vctk in d_vector mode, during 1M steps
then I added a specific voice to the dataset list and trained another 10K steps.

**Reply by Rico2000** [reply]
hey,
i im try to to train the model, but i have the problem that Training time is immense.  I use 4 V100 Gpus an it took like 5 hours for 1 epoch on btachsize 16. How did you achieve to train the model fast?

**Reply by WeberJulian** [reply]
Hey, that's definitely abnormal. First of all, you [can't](https://github.com/coqui-ai/TTS/blob/c63bb481e95bd4a1ff978947d8e3e6c0bfb4177f/TTS/tts/models/base_tts.py#L350) train multilingual on multiple GPUs since the language weighted sampler is not supported (except if you have a balanced dataset already). But yeah all training for our experiments were done on a single GPU (V100 or RTX 3090).

**Reply by Rico2000** [reply]
Hey,
oke i only try training on single language so multi gpu should work. do you have the config file for LibriTTS

**Reply by WeberJulian** [reply]
If it's not related to multilingual VITS you should probably open a new discussion/issue

**Reply by Rico2000** [reply]
oke i opened a new discussion : https://github.com/coqui-ai/TTS/discussions/1235

**Reply by makerjackie** [reply]
@WeberJulian 
can I train multilingual vits on multiple GPUs now? (on dev branch)
